ao-contracts/curve-bonded-token/src/handlers/token_manager.lua

```diff
@@ -0,0 +1,136 @@
+--- @type ReserveToken
+ReserveToken         = ReserveToken or nil
+--- @type ReserveBalance
+ReserveBalance       = ReserveBalance or '0'
+--- @type FundingGoal
+FundingGoal          = FundingGoal or '0'
+--- @type AllocationForLP
+AllocationForLP      = AllocationForLP or '0'
+--- @type AllocationForCreator
+AllocationForCreator = AllocationForCreator or '0'
+--- @type SupplyToSell
+SupplyToSell         = SupplyToSell or '0'
+--- @type MaxSupply
+MaxSupply            = MaxSupply or '0'
+--- @type Initialized
+Initialized          = Initialized or false
+--- @type ReachedFundingGoal
+ReachedFundingGoal   = ReachedFundingGoal or false
+--- @type LiquidityPool
+LiquidityPool        = LiquidityPool or nil
+--- @type Creator
+Creator              = Creator or nil
+
+---@type HandlerFunction
+function mod.initialize(msg)
+    assert(Initialized == false, "TokenManager already initialized")
+    assert(msg.Data ~= nil, "Data is required")
+
+    --- @type CBTMInitPayload
+    local initPayload = json.decode(msg.Data)
+
+    if (
+            validations.isInvalidInput(initPayload, 'object') or
+            validations.isInvalidInput(initPayload.repoToken, 'object') or
+            validations.isInvalidInput(initPayload.repoToken.tokenName, 'string') or
+            validations.isInvalidInput(initPayload.repoToken.tokenTicker, 'string') or
+            validations.isInvalidInput(initPayload.repoToken.denomination, 'string') or
+            validations.isInvalidInput(initPayload.repoToken.tokenImage, 'string') or
+            validations.isInvalidInput(initPayload.repoToken.processId, 'string') or
+            validations.isInvalidInput(initPayload.reserveToken, 'object') or
+            validations.isInvalidInput(initPayload.reserveToken.tokenName, 'string') or
+            validations.isInvalidInput(initPayload.reserveToken.tokenTicker, 'string') or
+            validations.isInvalidInput(initPayload.reserveToken.denomination, 'string') or
+            validations.isInvalidInput(initPayload.reserveToken.tokenImage, 'string') or
+            validations.isInvalidInput(initPayload.reserveToken.processId, 'string') or
+            validations.isInvalidInput(initPayload.fundingGoal, 'string') or
+            validations.isInvalidInput(initPayload.allocationForLP, 'string') or
+            validations.isInvalidInput(initPayload.allocationForCreator, 'string') or
+            validations.isInvalidInput(initPayload.maxSupply, 'string')
+        ) then
+        if msg.reply then
+            msg.reply({
+                Action = 'Initialize-Error',
+                Error = 'Invalid inputs supplied.'
+            })
+            return
+        else
+            ao.send({
+                Target = msg.From,
+                Action = 'Initialize-Error',
+                Error = 'Invalid inputs supplied.'
+            })
+        end
+    end
+
+    local lpAllocation = utils.udivide(utils.multiply(initPayload.maxSupply, "20"), "100")
```

Comment:
This should be calculating 20% but using `SubUnits` for `maxSupply`, because it's flooring on token integers, when it can be using flooring the % but for the subunits instead

In the edge case of creating a maxSupply of 4 tokens (with 12 as denomination), I can buy all 4 tokens and there won't be any supply reserved for the LP